### PR TITLE
Tests for Spicedham

### PR DESCRIPTION
**Pull request by iankronquist**
Tests for all methods in `spicedham/__init__.py`
@willkg r?

### CODE REVIEW

File: tests/test_spicedham.py [hunk 1]
```diff
@@ -1,67 +1,137 @@
...
+        plugin1.__name__ = "NotSqlAlchemyWrapper"
+        plugin1obj.classify.return_value = .75
+        plugin2 = Mock()
+        plugin2obj = Mock()
+        plugin2.return_value = plugin2obj
+        plugin2.__name__ = "StillNotSqlAlchemyWrapper"
+        plugin2obj = Mock()
+        plugin2.return_value = plugin2obj
+        plugin2obj.classify.return_value = None
+        mock_all_subclasses.return_value = [plugin0, plugin1, plugin2]
```
Comment (willkg) [inline tests/test_spicedham.py:28]: Looks like this gets repeated a lot. I have similarish messy patch code in fjord that I abstracted into a decorator.

https://github.com/mozilla/fjord/blob/master/fjord/translations/tests/test_gengo.py#L44

Those two decorators take arguments, so they're a bit more complicated than what I think you need here.

File: tests/test_spicedham.py [hunk 2]
```diff
@@ -1,67 +1,137 @@
...
+        self.assertTrue(plugin0obj.train.called)
 
-    def test_on_training_data(self, test_data_dir='corpus'):
-        self._test_all_files_in_dir(
-            os.path.join(test_data_dir, 'train', 'spam'), True)
-        self._test_all_files_in_dir(
-            os.path.join(test_data_dir, 'train', 'ham'), False)
+    @patch('spicedham.Spicedham.all_subclasses')
+    @patch('spicedham.Spicedham._load_backend')
+    def test_load_plugins(self, mock_all_subclasses, mock_load_backend):
```
Comment (willkg) [inline tests/test_spicedham.py]: This has the same name as the previous function.

**Comment by iankronquist** [comment]
@willkg r?

**Comment by willkg** [comment]
Can this also include a section in the README on how to run the tests?

**Comment by iankronquist** [comment]
@willkg done.

**Comment by willkg** [comment]
I don't see anything in that commit that adds a section to the README on how to run the tests.

**Comment by iankronquist** [comment]
Whoops, sorry. Forgot to push.

**Comment by willkg** [comment]
Looks good. I'm still getting one failure in `test_classifierbase`:

```
======================================================================
ERROR: Failure: ImportError (No module named test_classifierbase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/willkg/.virtualenvs/spicedham/lib/python2.7/site-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/willkg/.virtualenvs/spicedham/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/willkg/.virtualenvs/spicedham/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/willkg/mozilla/spicedham/tests/test_nonsensefilter.py", line 1, in <module>
    from tests.test_classifierbase import TestClassifierBase
ImportError: No module named test_classifierbase
```

I can spin that off into a new issue if you want. I have no idea what the issue is, so I don't know if this is easily fixed in this PR or not.

Otherwise this looks good. r+

**Comment by iankronquist** [comment]
Rebased, etc.

**Comment by iankronquist** [comment]
I screwed up the rebase so I'm going to have to fix something. See the last 2 commits.

**Pull request merged**